src/scenes/MerchStorePage/Redemption/useRedeemMerch.ts

```diff
@@ -0,0 +1,118 @@
+
+type Props = {
+  tokenIds: string[];
+  onSuccess: () => void;
+};
+
+export default function useRedeemMerch() {
+  const { address } = useAccount();
+
+  const [redeemMerch] = usePromisifiedMutation<useRedeemMerchMutation>(graphql`
+    mutation useRedeemMerchMutation($input: RedeemMerchInput!) @raw_response_type {
+      redeemMerch(input: $input) {
+        __typename
+        ... on RedeemMerchPayload {
+          discountCodes {
+            code
+            tokenId
+          }
+        }
+        ... on ErrInvalidInput {
+          __typename
+          message
+        }
+      }
+    }
+  `);
+
+  const { pushToast } = useToastActions();
+
+  return useCallback(
+    async ({ tokenIds, onSuccess }: Props) => {
+      try {
+        const signature = await signMessage({
+          message: `Gallery uses this cryptographic signature in place of a password: [${tokenIds.join(
+            ', '
+          )}]`,
+        });
+
+        const updater: SelectorStoreUpdater<useRedeemMerchMutation['response']> = (
+          store,
+          response
+        ) => {
+          if (response?.redeemMerch?.__typename === 'RedeemMerchPayload') {
+            const discountCodes = response?.redeemMerch?.discountCodes || [];
+
+            const root = store.get(`client:root:getMerchTokens(wallet:"${address}")`);
+
+            const tokens = root?.getLinkedRecords('tokens') || [];
+
+            // assign the discount code to the token
+            tokens.forEach((token) => {
+              const tokenId = token.getValue('tokenId');
+              const discountCode = discountCodes.find(
+                (discountCode) => discountCode.tokenId === tokenId
+              );
+
+              if (discountCode) {
+                token.setValue(discountCode.code, 'discountCode').setValue(true, 'redeemed');
+              }
+            });
+          }
+        };
+
+        const response = await redeemMerch({
+          variables: {
+            input: {
+              tokenIds,
+              address: {
+                address: (address as string).toLowerCase(),
+                chain: 'Ethereum',
+              },
+              walletType: 'EOA',
+              signature,
+            },
+          },
+          updater,
+        });
+
+        if (response?.redeemMerch?.__typename === 'ErrInvalidInput') {
+          pushToast({
+            message: 'Something went wrong',
+          });
+          return;
+        }
+
+        if (
+          response?.redeemMerch?.__typename === 'RedeemMerchPayload' &&
+          response?.redeemMerch?.discountCodes?.length === 0
+        ) {
+          pushToast({
+            message: 'Something went wrong',
+          });
+          return;
+        }
+
+        pushToast({
+          message: 'Successfully redeemed merch',
+        });
+
+        onSuccess();
+      } catch (error) {
+        pushToast({
+          message: 'Something went wrong',
```

Comment:
this copy feels a little cold. i'll push something up here

Reply:
Updated the copy with much better message